pkg/ingester/limiter.go

```diff
@@ -184,34 +185,43 @@ func (l *Limiter) convertGlobalToLocalLimit(userID string, globalLimit int) int
 		return 0
 	}
 
-	// Given we don't need a super accurate count (ie. when the ingesters
-	// topology changes) and we prefer to always be in favor of the tenant,
-	// we can use a per-ingester limit equal to:
-	// (global limit / number of ingesters) * replication factor
-	numIngesters := l.ring.HealthyInstancesCount()
-
-	// May happen because the number of ingesters is asynchronously updated.
-	// If happens, we just temporarily ignore the global limit.
-	if numIngesters == 0 {
-		return 0
+	zonesCount := l.getZonesCount()
+	var ingestersInZoneCount int
+	if zonesCount > 1 {
+		// In this case zone-aware replication is enabled, and ingestersInZoneCount is initially set to
+		// the total number of ingesters in the corresponding zone
+		ingestersInZoneCount = l.ring.InstancesInZoneCount()
+	} else {
+		// In this case zone-aware replication is disabled, and ingestersInZoneCount is initially set to
+		// the total number of ingesters
+		ingestersInZoneCount = l.ring.InstancesCount()
+	}
+	shardSize := l.getShardSize(userID)
+	// If shuffle sharding is enabled and the total number of ingesters in the zone is greater than the
+	// expected number of ingesters per sharded zone, then we should honor the latter because series/metadata
+	// cannot be written to more ingesters than that.
+	if shardSize > 0 {
+		ingestersInZoneCount = util_math.Min(ingestersInZoneCount, util.ShuffleShardExpectedInstancesPerZone(shardSize, zonesCount))
 	}
 
-	// If the number of available ingesters is greater than the tenant's shard
-	// size, then we should honor the shard size because series/metadata won't
-	// be written to more ingesters than it.
-	if shardSize := l.getShardSize(userID); shardSize > 0 {
-		// We use Min() to protect from the case the expected shard size is > available ingesters.
-		numIngesters = util_math.Min(numIngesters, util.ShuffleShardExpectedInstances(shardSize, l.getNumZones()))
+	// This may happen, for example when the total number of ingesters is asynchronously updated, or
+	// when zone-aware replication is enabled but ingesters in a zone have been scaled down.
+	// In those cases we ignore the global limit.
+	if ingestersInZoneCount == 0 {
+		return 0
 	}
 
-	return int((float64(globalLimit) / float64(numIngesters)) * float64(l.replicationFactor))
+	// Global limit is equally distributed among all the active zones.
+	// The portion of global limit related to each zone is then equally distributed
+	// among all the ingesters belonging to that zone.
+	return ((globalLimit * l.replicationFactor) / zonesCount) / ingestersInZoneCount
```

Comment:
Sorry for writing in italian. Let me translate it:

 I don’t think it’s correct to do the division operations in integer, because with each operation you lose precision. This is why we previously performed these operations in floating point and then converted the result to int.

Reply:
This was requested by @pstibrany. I will revert it so we ensure that we don't lose the precision

Reply:
translation:

> I don't think it's correct to do the division operations in integer, because with each operation you lose precision. This is why we first performed these operations in floating point and then converted the result to int.

Reply:
I suggested to either simplify that expression and remove floats, or add a test to show why floats are required. While I agree about loss of precision, I doubt it will have any meaningful effect here in practice.

Reply:
It's probably not meaningful because our limits are very high numbers, but I don't understand we shouldn't be correct.

Reply:
Max difference between using floats or not is `1 / ingestersInZoneCount`.

Reply:
Ok, then I just suggest to add a comment explaining why we do the int division, so that the next code reader won't ask themselves the same question.

Reply:
@duricanikolic I think Peter feels stronger than me on this, so I would get stick to what Peter suggested previously (int division). I would just add a comment explaining why it's not an issue ("Max difference between using floats or not is 1 / ingestersInZoneCount").

Reply:
A question for both @pracucci and @pstibrany: If we have a global limit 1000, replication factor 2 and number of zones 3, and let's say 1 ingester in that zone, should we return 666 or 667 as the result (1000 * 2 / 3)?

Reply:
Answer is: "doeasn't matter". There's no practical difference between the two. Which is Peter's point. I think int divisions are always floored (like casting float to int) so probably it's 666.

Reply:
> @duricanikolic I think Peter feels stronger than me on this, so I would get stick to what Peter suggested previously (int division).

No need to revert back.


> A question for both @pracucci and @pstibrany: If we have a global limit 1000, replication factor 2 and number of zones 3, and let's say 1 ingester in that zone, should we return 666 or 667 as the result (1000 * 2 / 3)?

Both are fine. Returning ceil would benefit user -- allow them to write one extra series per ingester.

Reply:
OK, so today I am going to test this PR on mimir-dev-09, and once it is done I will let you know, so we can merge this.